### PR TITLE
Toggle CloudWatch alarms actions during Data Refresh

### LINKED ISSUE #3647: Disable alarm notifications during ES index creation
## Problem

With each Data Refresh run, we got several notifications for alarms triggering its conditions due to rises in metrics like the API Response Times (AVG & P99) and Elasticsearch (ES) CPU usage. This is something expected since the index building puts pressure in the ES cluster, affecting its performance.

<br><br> 

But getting so many notifications every week is problematic because it makes maintainers get used to them, which may lead to miss when a real issue is happening and the situation worsens. On the other hand, it's annoying to get messages and stay alert for known and normal processes. 

## Description

One option to prevent excess of notifications for known behavior is to disable the alarms actions before ES index (re)creation, and activating them again when the process finishes. Currently this mainly occur with the Data Refresh but it would be useful for the new DAG for index creation (#3537).

The official Amazon Web Services Documentation repo has examples of using the [AWS Python SDK][aws-sdk] for this purpose, showing this is something possible and easily achievable.

[aws-sdk]: https://github.com/awsdocs/aws-doc-sdk-examples/tree/main/python/example_code/cloudwatch#code-examples

## Additional context

Related to #2344.

### PR DESCRIPTION
**Pull request by krysal**
## Fixes

Fixes #3647 by @krysal

## Description

This PR adds steps for disabling and re-enabling the actions of the most sensitive alarms when a Data Refresh process is running. I made the most basic changes to achieve this but maybe it can be improved to make these steps not required to continue the DAG execution and make sure that at the end the alarms actions are always re-enabled. Requesting @stacimc's review specifically for that.

We'll need to create AWS credentials for Airflow as well I believe.

## Testing Instructions

Set the AWS environment variables with the credentials of your AWS account. They can be found in the `~/.aws/credentials` file if you have configured the AWS CLI.

```
# .env file
AWS_ACCESS_KEY_ID=<your_key>
#AWS_SECRET_ACCESS_KEY=<your_secret>
#AWS_DEFAULT_REGION=us-east-1
```

🚨 This will modify production alarms actions 🚨 (if you have access to this credentials) but since a local data refresh finishes quickly and it's not destroying resources I thought this is acceptable if you want to see it working.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.
- [x] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Review by dhruvkb — APPROVED** [review]
LGTM, it's very cool that the DAG will automatically disable and re-enable the alarms!

**± 1 change:** The description of the `tasks` variable (lines 222 to 229), before calling `chain` needs updating.

File: catalog/dags/common/cloudwatch.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+
+class CloudWatchWrapper:
+    """Encapsulates Amazon CloudWatch functions"""
+
+    def __init__(self, cloudwatch_resource):
+        """:param cloudwatch_resource: A Boto3 CloudWatch resource."""
+        self.cloudwatch_resource = cloudwatch_resource
+
+    def enable_alarm_actions(self, alarm_name, enable):
```
[the remark this answers is not in the document]
Comment (krysal) [reply catalog/dags/common/cloudwatch.py:20]: I found it confusing first time I see it too, but then reframed it like the param being in imperative tone, so it dictates what the final state of the alarm should be it. Now it'd be strange to change it to be in other form IMO.

**Review by stacimc — CHANGES_REQUESTED** [review]
This is so cool -- really clever idea, and a welcome change since we investigate those false alarms _weekly_!

I have two blockers: first, the list of alarms being silenced. Can we instead only silence the ES Prod CPU Utilization? That one can 100% be ignored during the data refresh. For the response time I don't think instability in those should be silenced, especially since this part of the image data refresh takes _two days_, is generally running during the part of the week where the API is deployed, and those alarms are pretty critical.

Second and in response to your question in the PR description: we _can_ update this to ensure that `enable_alarms` runs even if the DAG fails during ingestion! To do that you need to add `trigger_rule=TriggerRule.ALL_DONE` to the `enable_alarms` task. This will cause that task to run as soon as all of its upstream tasks have completed, even if some of them failed.

The catch is that trigger rules only evaluate the status of a task's [immediately upstream tasks](https://stackoverflow.com/questions/66863896/airflow-trigger-rule-for-grandparent-tasks), so if the tasks were chained like this:

```
create_filtered_index >> enable_alarms >> promote_tasks
```

The `promote_tasks` will try to run as long as `enable_alarms` runs, even if `create_filtered_index` (or a further upstream task) failed. We can fix that by ensuring that both are upstream of the promotion steps:

```
tasks.append([enable_alarms, promote_tasks])
```

File: catalog/dags/data_refresh/data_refresh_task_factory.py [hunk 2]
```diff
@@ -201,7 +208,13 @@ def create_data_refresh_task_group(
...
             },
         )
-        tasks.append(delete_old_index)
+
+        enable_alarms = PythonOperator(
+            task_id="enable_cloudwatch_alarms",
+            python_callable=cloudwatch.enable_or_disable_alarms,
+            op_args=[True],
+        )
+        tasks.append([delete_old_index, enable_alarms])
```
[the remark this answers is not in the document]
Comment (krysal) [reply catalog/dags/data_refresh/data_refresh_task_factory.py]: Done as suggested in the main comment (in relation to dependencies).

File: catalog/dags/data_refresh/data_refresh_task_factory.py [hunk 3]
```diff
@@ -131,7 +132,13 @@ def create_data_refresh_task_group(
         generate_index_suffix = ingestion_server.generate_index_suffix.override(
             trigger_rule=TriggerRule.NONE_FAILED,
         )()
-        tasks.append(generate_index_suffix)
+
+        disable_alarms = PythonOperator(
+            task_id="disable_cloudwatch_alarms",
```
[the remark this answers is not in the document]
Comment (krysal) [reply catalog/dags/data_refresh/data_refresh_task_factory.py]: Done.

File: catalog/env.template [hunk 4]
```diff
@@ -88,9 +88,10 @@ AIRFLOW_PORT=9090
 LOADER_FILE_AGE=1
 # Contact email for any APIs
 CONTACT_EMAIL=openverse@wordpress.org
-# AWS/S3 configuration - does not need to be changed for development
-AWS_ACCESS_KEY=test_key
-AWS_SECRET_KEY=test_secret
+# AWS configuration - does not need to be changed for development
+AWS_ACCESS_KEY_ID=test_key
```
[the remark this answers is not in the document]
Comment (krysal) [reply catalog/env.template:92]: Well spotted! Updated.

File: catalog/dags/data_refresh/data_refresh_task_factory.py [hunk 5]
```diff
@@ -203,8 +210,14 @@ def create_data_refresh_task_group(
...
 
+        enable_alarms = PythonOperator(
+            task_id="enable_sensitive_cloudwatch_alarms",
+            python_callable=cloudwatch.enable_or_disable_alarms,
+            op_args=[True],
+            trigger_rule=TriggerRule.ALL_DONE,
+        )
         # Finally, promote the filtered index.
-        tasks.append(promote_filtered_index)
+        tasks.append([enable_alarms, promote_filtered_index])
```
Comment (stacimc) [inline catalog/dags/data_refresh/data_refresh_task_factory.py:220]: This also works, but still puts `enable_alarms` in between the promotion/deletion steps for the media index, and the promotion/deletion for the filtered index. I don't think that will cause any problems in terms of task execution, but it makes the flow of the DAG more confusing, and ideally we'd re-enable the alarm ASAP. I'd prefer to define this task earlier and then append it along with the first set of promote tasks on L199.

Comment (krysal) [reply catalog/dags/data_refresh/data_refresh_task_factory.py:220]: I see I chose the wrong promote task to pair it with. Gotcha, fixed.

**Review by stacimc — APPROVED** [review]
🥳 Tested all the failure states to be sure and the alarms are always re-enabled :) LGTM!

**Comment by krysal** [comment]
Thanks!

**Comment by AetherUnbound** [comment]
@krysal We run the data refresh DAG locally quite often for testing. Given that we don't have a Cloudwatch service available locally for testing, any local runs will immediately fail at the alarm step (I tried and confirmed this locally)
image

Do you mind adding in a mechanism, perhaps similar to our `SLACK_MESSAGE_OVERRIDE`, which checks the environment to make sure the step only happens in production _unless_ a Variable is explicitly set to `True`?

https://github.com/WordPress/openverse/blob/ee77ef5fe752fa5f4f89ca6d9c8f5af7fd4816ba/catalog/dags/common/slack.py#L304

**Pull request merged**